src/schema/error.rs

```diff
@@ -0,0 +1,276 @@
+        path: String,
+    },
+    /// Value does not match the required constant.
+    ConstMismatch {
+        expected: String,
+        actual: String,
+        path: String,
+    },
+    /// Value does not match any schema in a union (anyOf).
+    NoUnionMatch {
+        path: String,
+        errors: Vec<ValidationError>,
+    },
+    /// Invalid regex pattern in schema.
+    InvalidPattern { pattern: String, error: String },
+    /// Array item validation failed.
+    ArrayItemValidationFailed {
+        index: usize,
+        path: String,
+        error: Box<ValidationError>,
+    },
+    /// Object key is not a string.
+    NonStringKey { key_type: String, path: String },
+    /// Missing discriminator field in discriminated subobject.
+    MissingDiscriminator { discriminator: String, path: String },
+    /// Unknown discriminator value in discriminated subobject.
+    UnknownDiscriminatorValue {
+        discriminator: String,
+        value: String,
+        allowed_values: Vec<String>,
+        path: String,
+    },
+    /// Discriminated subobject validation failed.
+    DiscriminatedSubobjectValidationFailed {
+        discriminator: String,
+        value: String,
+        path: String,
+        error: Box<ValidationError>,
+    },
+}
+
+impl fmt::Display for ValidationError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ValidationError::TypeMismatch {
+                expected,
+                actual,
+                path,
+            } => {
+                write!(
+                    f,
+                    "Type mismatch at '{path}': expected {expected}, got {actual}"
+                )
+            }
+            ValidationError::OutOfRange {
+                value,
+                min,
+                max,
+                path,
+            } => {
+                let range_desc = match (min, max) {
+                    (Some(min), Some(max)) => format!("between {min} and {max}"),
+                    (Some(min), None) => format!("at least {min}"),
+                    (None, Some(max)) => format!("at most {max}"),
+                    (None, None) => "within valid range".to_string(),
+                };
+                write!(
+                    f,
+                    "Value {value} at '{path}' is out of range: must be {range_desc}"
+                )
+            }
+            ValidationError::LengthConstraint {
+                actual_length,
+                min_length,
+                max_length,
+                path,
+            } => {
+                let constraint_desc = match (min_length, max_length) {
+                    (Some(min), Some(max)) => format!("between {min} and {max} characters"),
+                    (Some(min), None) => format!("at least {min} characters"),
+                    (None, Some(max)) => format!("at most {max} characters"),
+                    (None, None) => "within valid length".to_string(),
```

Comment:
> (None, None) => "within valid length".to_string(), [](http://example.com/codeflow?start=19&length=51)

how does this case throw?

Reply:
Yes. That case can never be reached. However, the compiler is not smart enough to prove it and requires handling the (None, None) case.
Another thing to do here is to add an [`unreachable!'](https://doc.rust-lang.org/std/macro.unreachable.html). But that introduces a call to panic.

Reply:
I'm resolving this.
The good thing is that we are using strongly typed errors instead of [anyhow](https://docs.rs/anyhow/1.0.99/anyhow/) that is used in most places in Regorus.

When we attempt https://github.com/microsoft/regorus/issues/161, we can revisit the specific error messages.